src/Tools/ExternalAccess/Razor/RazorLanguageServerFactoryWrapper.cs

```diff
@@ -70,15 +80,43 @@ public IRazorLanguageServerTarget CreateLanguageServer(JsonRpc jsonRpc, IRazorCa
 
         private class RazorCapabilitiesProvider : ICapabilitiesProvider
         {
-            private readonly IRazorCapabilitiesProvider _razorCapabilitiesProvider;
+#pragma warning disable CS0618 // Type or member is obsolete
+            private readonly IRazorCapabilitiesProvider? _razorCapabilitiesProvider;
+#pragma warning restore CS0618 // Type or member is obsolete
+            private readonly IRazorTestCapabilitiesProvider? _razorTestCapabilitiesProvider;
 
+            public RazorCapabilitiesProvider(IRazorTestCapabilitiesProvider razorTestCapabilitiesProvider)
+            {
+                _razorTestCapabilitiesProvider = razorTestCapabilitiesProvider;
+            }
+
+#pragma warning disable CS0618 // Type or member is obsolete
             public RazorCapabilitiesProvider(IRazorCapabilitiesProvider razorCapabilitiesProvider)
+#pragma warning restore CS0618 // Type or member is obsolete
             {
                 _razorCapabilitiesProvider = razorCapabilitiesProvider;
             }
 
             public ServerCapabilities GetCapabilities(ClientCapabilities clientCapabilities)
-                => _razorCapabilitiesProvider.GetCapabilities(clientCapabilities);
+            {
+                if (_razorTestCapabilitiesProvider is not null)
+                {
+                    // To avoid exposing types from MS.VS.LanguageServer.Protocol types we serialize and deserialize the capabilities
+                    // so we can just pass string around. This is obviously not great for perf, but it is only used in Razor tests.
+                    var clientCapabilitiesJson = JsonConvert.SerializeObject(clientCapabilities);
```

Comment:
Might need this for proper serialization / deserialization
https://devdiv.visualstudio.com/DevDiv/_git/VSLanguageServerClient?path=/src/product/Protocol/LanguageServer.Protocol.Internal/Converters/VSInternalExtensionUtilities.cs&version=GBdevelop&line=24&lineEnd=25&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

We use it here for example - https://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/LanguageServer/AbstractInProcLanguageClient.cs,203

Reply:
Shouldn't need that, because I'm explicitly telling NewtonSoft that its a VSInternalServerCapabilities, which is the highest level type. The type converters are needed, AFAIK, for when you ask it to [de]serialize a lower level type, but don't want to lose the info from the higher types.

Also, this is only used by tests, our scenario is _super_ simple. We only set `SupportsDiagnosticsRequests`, and always serialize a VSInternalServerCapabilities: https://github.com/dotnet/razor/blob/main/src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/SingleServerDelegatingEndpointTestBase.cs#L75